src/PresetsLoader.cpp

```diff
@@ -0,0 +1,175 @@
+/*
+ *  Copyright (C) 2005-2021 Team Kodi (https://kodi.tv)
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSE.md for more information.
+ */
+
+#include "PresetsLoader.h"
+
+#include <json/json.h>
+#include <kodi/Filesystem.h>
+#include <kodi/General.h>
+#include <kodi/tools/StringUtils.h>
+
+bool CPresetLoader::Load(const std::string& path)
+{
+  bool ret = false;
+  char* ptr = nullptr;
+
+  try
+  {
+    kodi::vfs::CFile file;
+    if (!file.OpenFile(path))
+    {
+      throw kodi::tools::StringUtils::Format("Failed to open shader list file %s", path.c_str());
+    }
+
+    ssize_t length = file.GetLength();
+    if (length <= 0)
+    {
+      throw kodi::tools::StringUtils::Format("Shader list file %s seems empty and not usable",
+                                             path.c_str());
+    }
+
+    ptr = new char[length + 1]{0};
+    ssize_t pos = 0;
+    do
+    {
+      ssize_t ret = file.Read(ptr, length);
+      if (ret < 0)
+      {
+        throw kodi::tools::StringUtils::Format("Failed to read shader list file %s", path.c_str());
+      }
+      pos += ret;
+    } while (pos < length);
+
+    try
+    {
+      JSONCPP_STRING err;
+      Json::Value root;
+      Json::CharReaderBuilder builder;
+      const std::unique_ptr<Json::CharReader> reader(builder.newCharReader());
+      if (!reader->parse(ptr, ptr + length, &root, &err))
+      {
+        throw kodi::tools::StringUtils::Format(
+            "Failed to parse json within shader list file %s (ERROR: %s)", path.c_str(),
+            err.c_str());
+        ;
+      }
+
+      for (unsigned int index = 0; index < root["presets"].size(); index++)
+      {
+        const auto& entry = root["presets"][index];
+        Preset preset;
+
+        // Check optional value about support limitation added
```

Comment:
This comment is worded a bit creatively, can you explain what it means? I'll help get it understandable for everyone

Reply:
Yes sounds really bad. Maybe here a bit better: `Check the optional JSON value at the end of the preset, which sets possible limits, e.g. support only to GL ("gl_only").` 

Thought make there below the code about checks and thought to hold the description text a bit more general for case another checks comes to it in future.

Reply:
OK how about

```c++
// Check the optional JSON value at the end of the preset, which sets possible limits, e.g. supporting only GL ("gl_only")

Reply:
Thanks, has changed to it.